# import/esm-extensions

ðŸ’¼ This rule is enabled in the `recommended-esm` config.

ðŸ”§ðŸ’¡ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix) and manually fixable by [editor suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).

<!-- end auto-generated rule header -->

Check for full path imports (as required by ESM modules).

This rule will flag any relative import paths that do not resolve to a file.

```js
import foo from './my/path/index.js';
```

## Rule Details

The below examples are assuming you have the following directory contents:

```bash
src
 â”œ ExampleFile.js
 â”œ core
 â”‚  â”œ index.js
 â”‚  â”œ importType.js
 â”‚  â”œ packagePath.js
 â”‚  â”œ staticRequire.js
 â”œ docsUrl.js
 â”œ importDeclaration.js
 â”œ index.js

```

and the below code exists within `src/ExampleFile.js`

### Fail

```js
import myModule from '.'; // cjs would import src/index.js
import myModule from './'; // cjs would import src/index.js
import myModule from './index'; // cjs would import src/index.js
import myModule from './index'; // cjs would import src/index.js
import core from './core'; // cjs would import src/core/index.js
import core from './core/index'; // cjs would import src/core/index.js
import packagePath from './core/packagePath'; // cjs would import src/core/packagePath.js
```

### Pass

```js
// This is what the previous failing imports would be fixed to when using `--fix`
import myModule from './index.js'; // This is what the previous failing imports would be fixed to
import core from './core/index.js'; // This is what the previous failing imports would be fixed to
import packagePath from './core/packagePath.js'; // cjs would import src/core/packagePath.js
```

## When Not To Use It

If you're fully migrated to ESM, you probably have other rules to handle this, but this rule could still be used in place of those to flag invalid ESM imports.
